node/src/components/linear_chain.rs

```diff
@@ -185,49 +250,43 @@ where
-                // If new, gossip and store.
-                if signature_known {
-                    self.remove_from_pending_fs(&*fs);
-                    Effects::new()
-                } else {
-                    let mut effects = effect_builder
-                        .announce_finality_signature(fs.clone())
-                        .ignore();
-                    if let Some(signature) = self.remove_from_pending_fs(&*fs) {
-                        if signature.is_local() {
-                            let message = Message::FinalitySignature(fs.clone());
-                            effects.extend(effect_builder.broadcast_message(message).ignore());
-                        }
-                    };
-                    signatures.insert_proof(fs.public_key, fs.signature);
-                    // Cache the results in case we receive the same finality signature before we
-                    // manage to store it in the database.
-                    self.cache_signatures(*signatures.clone());
-                    debug!(hash=%signatures.block_hash, "storing finality signatures");
-                    effects.extend(
-                        effect_builder
-                            .put_signatures_to_storage(*signatures)
-                            .ignore(),
-                    );
-                    effects
-                }
+            Event::IsBonded(Some(mut known_signatures), fs, true) => {
+                // New finality signature from a bonded validator.
+                known_signatures.insert_proof(fs.public_key, fs.signature);
+                // Cache the results in case we receive the same finality signature before we
+                // manage to store it in the database.
+                self.linear_chain_state
+                    .cache_signatures(*known_signatures.clone());
+                debug!(hash=%known_signatures.block_hash, "storing finality signatures");
+                // Announce new finality signatures for other components to pick up.
+                let mut effects = effect_builder
+                    .announce_finality_signature(fs.clone())
+                    .ignore();
+                if let Some(signature) = self.linear_chain_state.remove_from_pending_fs(&*fs) {
+                    // This shouldn't return `None` as we added the `fs` to the pending collection
+                    // when we received it. If it _is_ `None` then a concurrent
+                    // flow must have already removed it. If it's a signature
+                    // created by this node, gossip it.
+                    if signature.is_local() {
+                        let message = Message::FinalitySignature(fs.clone());
+                        effects.extend(effect_builder.broadcast_message(message).ignore());
+                    }
+                };
+                effects.extend(
+                    effect_builder
+                        .put_signatures_to_storage(*known_signatures)
```

Comment:
I wonder whether it can happen that the put events get handled in the wrong order.
I don't think it can, currently, but it's also not something components should rely on, in general.

E.g. if you got a signature A from Alice, put it in the cache, and made the put event for `[A]`. Then you got B from Bob, and made the put event for `[A, B]`. Then the latter even is handled, and then the former. The state would now contain only `[A]`.

(Not something for this PR, but maybe to revisit later?)

Reply:
Storage also combined the signatures together – it adds the new data to the old one. I made sure that's the case when that piece was written.